heltour/tournament/views.py

```diff
@@ -615,62 +614,61 @@ def view(self, username):
 
-class RegisterView(LeagueView):
+class RegisterView(LoginRequiredMixin, LeagueView):
+
     def view(self, post=False):
-        if self.season is not None and self.season.registration_open:
-            reg_season = self.season
-        else:
-            reg_season = Season.objects.filter(league=self.league, registration_open=True).order_by('-start_date').first()
+        reg_season = Season.get_registration_season(self.league, self.season)
         if reg_season is None:
             return self.render('tournament/registration_closed.html', {})
-        if post:
-            form = RegistrationForm(self.request.POST, season=reg_season)
-            if form.is_valid():
-                with reversion.create_revision():
-                    reversion.set_comment('Submitted registration.')
-                    form.save()
-                # send registration received email
-                subject = render_to_string('tournament/emails/registration_received_subject.txt', {'reg': form.instance})
-                msg_plain = render_to_string('tournament/emails/registration_received.txt', {'reg': form.instance})
-                msg_html = render_to_string('tournament/emails/registration_received.html', {'reg': form.instance})
-                try:
-                    send_mail(
-                        subject,
-                        msg_plain,
-                        settings.DEFAULT_FROM_EMAIL,
-                        [form.cleaned_data['email']],
-                        html_message=msg_html,
-                    )
-                except SMTPException:
-                    logger.exception('A confirmation email could not be sent.')
-
-                self.request.session['reg_email'] = form.cleaned_data['email']
-                return redirect(leagueurl('registration_success', league_tag=self.league.tag, season_tag=self.season.tag))
-        else:
-            form = RegistrationForm(season=reg_season)
-            if self.request.user.is_authenticated():
-                player = Player.get_or_create(self.request.user.username)
-                form.fields['lichess_username'].initial = player.lichess_username
-                form.fields['email'].initial = player.email
-                form.fields['classical_rating'].initial = player.rating_for(reg_season.league)
-                form.fields['has_played_20_games'].initial = not player.provisional_for(reg_season.league)
-                form.fields['already_in_slack_group'].initial = player.slack_user_id != ''
 
-        context = {
-            'form': form,
-            'registration_season': reg_season
-        }
-        return self.render('tournament/register.html', context)
+        with cache.lock(f'update_create_registration-{self.request.user.id}-{reg_season.id}'):
+            instance = Registration.get_registration(self.request.user, reg_season)
+            if post:
+                form = RegistrationForm(self.request.POST, instance=instance, season=reg_season)
```

Comment:
The registration status should be reset to "pending" before saving (e.g. the user registers, is approved, withdraws, and registers again)

Reply:
Or, perhaps better, have it create a new registration in this case.

Reply:
I'm worried about what happens in the case of approved members who change their registration.  Their status would go back to pending and mods would have to re-approve them.  Is this what we want every time?  What code is triggered when the mods approve someone?

Reply:
If an already participating player is approved (or rejected) again nothing crazy happens. It would take a deliberate action to move them off a team.
https://github.com/cyanfish/heltour/blob/master/heltour/tournament/workflows.py#L367

Reply:
Part of the point of this PR is to prevent duplicate registrations. Basically, once a player registers, they can't register again, they can only change their registration in some well defined ways

But getting a bunch of duplicate registrations is annoying to deal with

Reply:
Or am I mistaken in what this line is about

Reply:
There needs to be some way to re-register after approval. If a user's registration is pending this will still prevent duplicates. It will also prevent re-registration in case of rejection.

To narrow down this other case (re-registration after approval), we can also block registration changes if the season player is active without any pending withdrawals.

Reply:
Soo I found lines 133-139 in forms.py.  status is already automatically set to pending every time the registration form is saved so I think this is already done.

Reply:
Ok, yes it does work as expected.  Status is already reset to pending every time the form is saved.